### PR TITLE
Better support for displaced entrances and exits

### PR DESCRIPTION
**Pull request by Broxzier**
Rides have a list of XY locations for the entrances and exits of every station, and a list of station heights which basically acts as the Z axis. At many places the game assumes that the entrances are there, making them not work properly when they have been moved. As a result, building a working queue from the entrance is only possible by copying over a path from the actual station, and peeps will try to navigate to the original location.

This PR makes detached ride entrances usable by making the path tool search for them on the known tile when placing a path. Instead of checking for the expected height, it loops over the elements on tile XY and checks for the ride index instead (which it didn't do before, so it happens you have two entrances at the same tile, it would always choose the first in the list, even if it's from another ride!).

To make entrances and exits usable that have been moved, I've added a button to the Tile Inspector which can update the location that the ride has saved.

This PR only takes care of the assumption for when building queues, but there are many more (example: mechanics don't get called when the ride exit is not where the ride thinks it is). I'd like to leave those for later, preferably after the next release.

### CODE REVIEW

File: src/openrct2/world/Footpath.cpp [hunk 1]
```diff
@@ -1598,27 +1598,29 @@ void footpath_queue_chain_push(uint8 rideIndex)
  */
 void footpath_update_queue_chains()
 {
-    for (uint8 *queueChainPtr = _footpathQueueChain; queueChainPtr < _footpathQueueChainNext; queueChainPtr++) {
-        uint8 rideIndex = *queueChainPtr;
-        Ride *ride = get_ride(rideIndex);
-        if (ride->type == RIDE_TYPE_NULL) {
+    for (uint8 * queueChainPtr = _footpathQueueChain; queueChainPtr < _footpathQueueChainNext; queueChainPtr++)
```
Comment (janisozaur) [inline src/openrct2/world/Footpath.cpp]: This should use range-based `for` and there's little reason for `queueChainPtr` and `rideIndex` to coexist,

Comment (janisozaur) [reply src/openrct2/world/Footpath.cpp]: Oh, it doesn't loop through all the elements, but only up to a pointer somewhere into this array. Nevermind then.

File: src/openrct2/world/TileInspector.cpp [hunk 2]
```diff
@@ -587,6 +587,42 @@ sint32 tile_inspector_path_toggle_edge(sint32 x, sint32 y, sint32 elementIndex,
     return 0;
 }
 
+sint32 tile_inspector_entrance_make_usable(sint32 x, sint32 y, sint32 elementIndex, sint32 flags)
+{
+    rct_tile_element * const entranceElement = map_get_nth_element_at(x, y, elementIndex);
+
+    if (!entranceElement || tile_element_get_type(entranceElement) != TILE_ELEMENT_TYPE_ENTRANCE)
```
Comment (rwjuk) [inline src/openrct2/world/TileInspector.cpp]: `if (entranceElement == nullptr || ...`

File: src/openrct2/world/TileInspector.cpp [hunk 3]
```diff
@@ -587,6 +587,42 @@ sint32 tile_inspector_path_toggle_edge(sint32 x, sint32 y, sint32 elementIndex,
...
+    {
+        return MONEY32_UNDEFINED;
+    }
+
+    if (flags & GAME_COMMAND_FLAG_APPLY)
+    {
+        Ride * ride         = get_ride(entranceElement->properties.entrance.ride_index);
```
Comment (rwjuk) [inline src/openrct2/world/TileInspector.cpp]: Should check for `nullptr`.

**Comment by rwjuk** [comment]
Left a couple of comments regarding null checks, otherwise LGTM (will need a rebase).

File: src/openrct2/world/TileInspector.cpp [hunk 4]
```diff
@@ -1,23 +1,23 @@
...
-* For more information, visit https://github.com/OpenRCT2/OpenRCT2
-*
-* OpenRCT2 is free software: you can redistribute it and/or modify
-* it under the terms of the GNU General Public License as published by
-* the Free Software Foundation, either version 3 of the License, or
-* (at your option) any later version.
-*
-* A full copy of the GNU General Public License can be found in licence.txt
-*****************************************************************************/
+ * OpenRCT2, an open source clone of Roller Coaster Tycoon 2.
```
Comment (Gymnasiast) [inline src/openrct2/world/TileInspector.cpp]: ?

File: src/openrct2/world/TileInspector.cpp [hunk 5]
```diff
@@ -89,9 +89,7 @@ sint32 tile_inspector_insert_corrupt_at(sint32 x, sint32 y, sint16 elementIndex,
 {
     // Make sure there is enough space for the new element
     if (!map_check_free_elements_and_reorganise(1))
-    {
```
Comment (Gymnasiast) [inline src/openrct2/world/TileInspector.cpp]: Unnecessary change.

**Comment by Gymnasiast** [comment]
> At many places the game assumes that the entrances are there, making them not work properly when they have been moved

Not true. When loading a save, the game checks if all entrances and exits are actually at the set locations, and updates those locations if they are not there. This can only fail when there are multiple entrance/exit buildings.

File: src/openrct2/world/TileInspector.cpp (hunk 4 above)
Comment (rwjuk) [reply src/openrct2/world/TileInspector.cpp]: I think this is just what clang-format does - we should probably tweak it not to do it.

Comment (Broxzier) [reply src/openrct2/world/TileInspector.cpp]: clang-format screws up the copyright notice at the top, and adds those spaces. After comparing it with other files I noticed that they all have one space before the lines, so I left that change in.

File: src/openrct2/world/TileInspector.cpp (hunk 5 above)
Comment (Broxzier) [reply src/openrct2/world/TileInspector.cpp]: It's more consistent, though.

**Comment by Broxzier** [comment]
> Not true. When loading a save, the game checks if all entrances and exits are actually at the set locations, and updates those locations if they are not there. This can only fail when there are multiple entrance/exit buildings.

Most players will try to build a path from an entrance or exit right after placing it instead of reloading the saved game first. With this PR they are usable right away.

**Comment by Gymnasiast** [comment]
This does not update station_heights, which might cause problems with pathfinding. On the other hand, this value is shared by the entrance and exit, so it's hard to say whether it should update that value or not. Basically, any scenario where the entrance and exit are not at the some level will cause trouble.

**Comment by Broxzier** [comment]
The exit needs to match the station height for a mechanic to find fix the ride, but guests don't care about the actual entrance height apparently. They're able to find the entrance when it's higher up.

**Comment by Gymnasiast** [comment]
Hm, so perhaps a good idea to only update it when moving the exit?

**Comment by Broxzier** [comment]
I thought about doing that, but then I want it to be a temporary fix, until the mechanic AI is improved. I'd like to get rid of all assumptions of the station height matching the entrance and exit heights. Something like this: https://gist.github.com/Broxzier/9fec1b15ddc23f59ddc24a1d69688b2d
But I'll have to look into this in more detail. I think that better fits in its own PR (which then should remove the temporary fix).

**Comment by Gymnasiast** [comment]
I think that's a good idea, yes. Can you put in the temporary hack with a TODO note?

**Comment by Broxzier** [comment]
Done so, and rebased. I also did a quick test to make sure it actually works.
